activemodel/lib/active_model/type/helpers/numeric.rb

```diff
@@ -25,6 +25,8 @@ def cast(value)
         end
 
         def changed?(old_value, _new_value, new_value_before_type_cast) # :nodoc:
+          return false if old_value.try(:nan?) && new_value_before_type_cast.try(:nan?)
```

Comment:
I worry about the performance overhead of this. `try` isn't cheap.

We can be a bit faster by using `.equal?`:

```ruby
require 'active_support/all'
require 'benchmark/ips'

old_value = 43
new_value = 44.0

Benchmark.ips do |x|
  x.report('try') do
    old_value.try(:nan?) && new_value.try(:nan?)
  end
  x.report('equal') do
    (Float::NAN.equal?(old_value) && Float::NAN.equal?(new_value)) ||
      (BigDecimal::NAN.equal?(old_value) && BigDecimal::NAN.equal?(old_value))
  end
  x.compare!
end
```

```
Calculating -------------------------------------
                 try      3.562M (±14.9%) i/s -     17.297M in   5.016700s
               equal      6.578M (±13.2%) i/s -     32.403M in   5.060069s

Comparison:
               equal:  6577869.5 i/s
                 try:  3561709.2 i/s - 1.85x  (± 0.00) slower
```

Reply:
Also in your version changing a value from `Float::NAN` to `BigDecimal::NAN` wouldn't mark the attribute as dirty. Using `.equal?` would.

Reply:
Alternatively we could have a single `equal?` in https://github.com/rails/rails/blob/212077a972404f19aa520a7792d2b3355f2de741/activemodel/lib/active_model/type/value.rb#L72-L74

Reply:
```ruby
require 'active_support/all'
require 'benchmark/ips'

old_value = 43
new_value = 44.0

Benchmark.ips do |x|
  x.report('try') do
    old_value.try(:nan?) && new_value.try(:nan?)
  end
  x.report('equal') do
    old_value.equal?(new_value)
  end
  x.compare!
end
```

```
Calculating -------------------------------------
                 try      3.922M (± 3.3%) i/s -     19.648M in   5.015799s
               equal     11.098M (± 1.6%) i/s -     55.764M in   5.025777s

Comparison:
               equal: 11098414.5 i/s
                 try:  3922009.2 i/s - 2.83x  (± 0.00) slower
```

So minimal overhead, minimal extra code. Maybe worth a code comment though as it might throw readers off.

Reply:
@byroot `equal?` won't work because it checks if both values are the same object, for example this evaluates to false
```ruby
old_value = "abc"
new_value = "abc"
old_value.equal?(new_value) # false
```

I did a similar solution to that that you have purposed in your first comment which is faster than `try` and covers all scenarios

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "activesupport"
  gem "benchmark-ips"
end

require 'active_support/all'
require 'benchmark/ips'

old_value = 55.0
new_value = 54.0

Benchmark.ips do |x|
  x.report('try') do
    old_value.try(:nan?) && new_value.try(:nan?)
  end
  x.report('unchanged_nan?') do
    ((old_value.is_a?(Float) || old_value.is_a?(BigDecimal)) &&
      (new_value.is_a?(Float) || new_value.is_a?(BigDecimal))) && (old_value.nan? && new_value.nan?) && old_value.equal?(new_value)
  end
  x.compare!
end
```
```
Calculating -------------------------------------
                 try      3.876M (± 3.8%) i/s -     19.350M in   5.000831s
      unchanged_nan?     10.996M (± 0.9%) i/s -     55.464M in   5.044266s

Comparison:
      unchanged_nan?: 10996335.8 i/s
                 try:  3875589.3 i/s - 2.84x  (± 0.00) slower
```

Reply:
Damn it, I assume `NaN` was a singleton like `nil` & such:

```
>> 0.0/0
=> NaN
>> (0.0/0).object_id
=> 180
>> (0.0/0).object_id
=> 200
```

Unfortunately it's not :/